File: sdk7demo/ios/sdk7demoTests/DecodingTests.swift
//
//  DecodingTests.swift
//  sdk7demoTests
//
//  Created by Gábor Vass on 17/02/2021.
//

import XCTest
import GetSocialSDK

class DecodingTests: XCTestCase {

  func test_action() {
    self.execute(folder: "action", onDecode: { (result: Action?) in
      XCTAssertNotNil(result)
    })
  }

  func test_activitiesquery() {
    self.execute(folder: "activitiesquery", onDecode: { (result: ActivitiesQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_activitybutton() {
    self.execute(folder: "activitybutton", onDecode: { (result: ActivityButton?) in
      XCTAssertNotNil(result)
    })
  }

  func test_activitycontent() {
    self.execute(folder: "activitycontent", onDecode: { (result: ActivityContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_addgroupmembersquery() {
    self.execute(folder: "addgroupmembersquery", onDecode: { (result: UpdateGroupMembersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_announcementsquery() {
    self.execute(folder: "announcementsquery", onDecode: { (result: AnnouncementsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_chatid() {
    self.execute(folder: "chatid", onDecode: { (result: ChatId?) in
      XCTAssertNotNil(result)
    })
  }

  func test_chatmessagecontent() {
    self.execute(folder: "chatmessagecontent", onDecode: { (result: ChatMessageContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_chatmessagespagingquery() {
    self.execute(folder: "chatmessagespagingquery", onDecode: { (result: ChatMessagesPagingQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_chatmessagesquery() {
    self.execute(folder: "chatmessagesquery", onDecode: { (result: ChatMessagesQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_followersquery() {
    self.execute(folder: "followersquery", onDecode: { (result: FollowersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_followquery() {
    self.execute(folder: "followquery", onDecode: { (result: FollowQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_friendsquery() {
    self.execute(folder: "friendsquery", onDecode: { (result: FriendsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_groupcontent() {
    self.execute(folder: "groupcontent", onDecode: { (result: GroupContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_groupsquery() {
    self.execute(folder: "groupsquery", onDecode: { (result: GroupsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_identity() {
    self.execute(folder: "identity", onDecode: { (result: Identity?) in
      XCTAssertNotNil(result)
    })
  }

  func test_joingroupquery() {
    self.execute(folder: "joingroupquery", onDecode: { (result: UpdateGroupMembersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_mediaattachment() {
    self.execute(folder: "mediaattachment", onDecode: { (result: MediaAttachment?) in
      XCTAssertNotNil(result)
    })
  }

  func test_membersquery() {
    self.execute(folder: "membersquery", onDecode: { (result: MembersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_polloptioncontent() {
    self.execute(folder: "polloptioncontent", onDecode: { (result: PollOptionContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_pollcontent() {
    self.execute(folder: "pollcontent", onDecode: { (result: PollContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_postactivitytarget() {
    self.execute(folder: "postactivitytarget", onDecode: { (result: PostActivityTarget?) in
      XCTAssertNotNil(result)
    })
  }

  // TODO: fix purchase data
  func atest_purchaseData() {
    self.execute(folder: "purchasedata", onDecode: { (result: PurchaseData?) in
      //XCTAssertNotNil(result)
    })
  }

  func test_reactionsquery() {
    self.execute(folder: "reactionsquery", onDecode: { (result: ReactionsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_removeactivitiesquery() {
    self.execute(folder: "removeactivitiesquery", onDecode: { (result: RemoveActivitiesQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_removegroupmembersquery() {
    self.execute(folder: "removegroupmembersquery", onDecode: { (result: RemoveGroupMembersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_tagsquery() {
    self.execute(folder: "tagsquery", onDecode: { (result: TagsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_topicsquery() {
    self.execute(folder: "topicsquery", onDecode: { (result: TopicsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_usersquery() {
    self.execute(folder: "usersquery", onDecode: { (result: UsersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_votesquery() {
    self.execute(folder: "votesquery", onDecode: { (result: VotesQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_userid() {
    self.execute(folder: "userid", onDecode: { (result: UserId?) in
      XCTAssertNotNil(result)
    })
  }

  func test_useridlist() {
    self.execute(folder: "useridlist", onDecode: { (result: UserIdList?) in
      XCTAssertNotNil(result)
    })
  }

  func test_userupdate() {
    self.execute(folder: "userupdate", onDecode: { (result: UserUpdate?) in
      XCTAssertNotNil(result)
    })
  }

  func test_invitecontent() {
    self.execute(folder: "invitecontent", onDecode: { (result: InviteContent?) in
      XCTAssertNotNil(result)
    })
  }

  func test_notificationbadge() {
    self.execute(folder: "notificationbadge", onDecode: { (result: NotificationBadge?) in
      XCTAssertNotNil(result)
    })
  }

  func test_notificationcustomization() {
    self.execute(folder: "notificationcustomization", onDecode: { (result: NotificationCustomization?) in
      XCTAssertNotNil(result)
    })
  }

  func test_notificationsquery() {
    self.execute(folder: "notificationsquery", onDecode: { (result: NotificationsQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_sendnotificationtarget() {
    self.execute(folder: "sendnotificationtarget", onDecode: { (result: SendNotificationTarget?) in
      XCTAssertNotNil(result)
    })
  }

  func test_referralusersquery() {
    self.execute(folder: "referralusersquery", onDecode: { (result: ReferralUsersQuery?) in
      XCTAssertNotNil(result)
    })
  }

  func test_promocodecontent() {
    self.execute(folder: "promocodecontent", onDecode: { (result: PromoCodeContent?) in
      XCTAssertNotNil(result)
    })
  }

}

File: sdk7demo/getsocial-react-native-sdk/android/src/test/java/im/getsocial/rn/DecodingTests.java
package im.getsocial.rn;

import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;

import im.getsocial.sdk.actions.Action;
import im.getsocial.sdk.communities.ActivitiesQuery;
import im.getsocial.sdk.communities.ActivityButton;
import im.getsocial.sdk.communities.ActivityContent;
import im.getsocial.sdk.communities.AddGroupMembersQuery;
import im.getsocial.sdk.communities.AnnouncementsQuery;
import im.getsocial.sdk.communities.ChatId;
import im.getsocial.sdk.communities.ChatMessageContent;
import im.getsocial.sdk.communities.ChatMessagesPagingQuery;
import im.getsocial.sdk.communities.ChatMessagesQuery;
import im.getsocial.sdk.communities.FollowQuery;
import im.getsocial.sdk.communities.FollowersQuery;
import im.getsocial.sdk.communities.FriendsQuery;
import im.getsocial.sdk.communities.GroupContent;
import im.getsocial.sdk.communities.GroupsQuery;
import im.getsocial.sdk.communities.Identity;
import im.getsocial.sdk.communities.JoinGroupQuery;
import im.getsocial.sdk.communities.MembersQuery;
import im.getsocial.sdk.communities.PollContent;
import im.getsocial.sdk.communities.PollOptionContent;
import im.getsocial.sdk.communities.PostActivityTarget;
import im.getsocial.sdk.communities.ReactionsQuery;
import im.getsocial.sdk.communities.RemoveActivitiesQuery;
import im.getsocial.sdk.communities.RemoveGroupMembersQuery;
import im.getsocial.sdk.communities.TagsQuery;
import im.getsocial.sdk.communities.TopicsQuery;
import im.getsocial.sdk.communities.UserId;
import im.getsocial.sdk.communities.UserIdList;
import im.getsocial.sdk.communities.UserUpdate;
import im.getsocial.sdk.communities.UsersQuery;
import im.getsocial.sdk.communities.VotesQuery;
import im.getsocial.sdk.iap.PurchaseData;
import im.getsocial.sdk.invites.InviteContent;
import im.getsocial.sdk.invites.ReferralUsersQuery;
import im.getsocial.sdk.json.GetSocialJson;
import im.getsocial.sdk.media.MediaAttachment;
import im.getsocial.sdk.notifications.NotificationContent;
import im.getsocial.sdk.notifications.NotificationCustomization;
import im.getsocial.sdk.notifications.NotificationsQuery;
import im.getsocial.sdk.notifications.SendNotificationTarget;
import im.getsocial.sdk.promocodes.PromoCodeContent;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class DecodingTests {

    @Test
    public void test_Action() throws FileNotFoundException {
        assertTrue(execute("action", Action.class));
    }

    @Test
    public void test_ActivitiesQuery() throws FileNotFoundException {
        assertTrue(execute("activitiesquery", ActivitiesQuery.class));
    }

    @Test
    public void test_ActivityButton() throws FileNotFoundException {
        assertTrue(execute("activitybutton", ActivityButton.class));
    }

    @Test
    public void test_ActivityContent() throws FileNotFoundException {
        assertTrue(execute("activitycontent", ActivityContent.class));
    }

    @Test
    public void test_AddGroupMembersQuery() throws FileNotFoundException {
        assertTrue(execute("addgroupmembersquery", AddGroupMembersQuery.class));
    }

    @Test
    public void test_AnnouncementsQuery() throws FileNotFoundException {
        assertTrue(execute("announcementsquery", AnnouncementsQuery.class));
    }
    @Test
    public void test_ChatId() throws FileNotFoundException {
        assertTrue(execute("chatid", ChatId.class));
    }
    @Test
    public void test_ChatMessageContent() throws FileNotFoundException {
        assertTrue(execute("chatmessagecontent", ChatMessageContent.class));
    }
    @Test
    public void test_ChatMessagesPagingQuery() throws FileNotFoundException {
        assertTrue(execute("chatmessagespagingquery", ChatMessagesPagingQuery.class));
    }
    @Test
    public void test_ChatMessagesQuery() throws FileNotFoundException {
        assertTrue(execute("chatmessagesquery", ChatMessagesQuery.class));
    }
    @Test
    public void test_FollowersQuery() throws FileNotFoundException {
        assertTrue(execute("followersquery", FollowersQuery.class));
    }
    @Test
    public void test_FollowQuery() throws FileNotFoundException {
        assertTrue(execute("followquery", FollowQuery.class));
    }
    @Test
    public void test_FriendsQuery() throws FileNotFoundException {
        assertTrue(execute("friendsquery", FriendsQuery.class));
    }
    @Test
    public void test_GroupContent() throws FileNotFoundException {
        assertTrue(execute("groupcontent", GroupContent.class));
    }
    @Test
    public void test_GroupsQuery() throws FileNotFoundException {
        assertTrue(execute("groupsquery", GroupsQuery.class));
    }
    @Test
    public void test_Identity() throws FileNotFoundException {
        assertTrue(execute("identity", Identity.class));
    }
    @Test
    public void test_JoinGroupQuery() throws FileNotFoundException {
        assertTrue(execute("joingroupquery", JoinGroupQuery.class));
    }
    @Test
    public void test_MediaAttachment() throws FileNotFoundException {
        assertTrue(execute("mediaattachment", MediaAttachment.class));
    }
    @Test
    public void test_MembersQuery() throws FileNotFoundException {
        assertTrue(execute("membersquery", MembersQuery.class));
    }

    @Test
    public void test_PollContent() throws FileNotFoundException {
        assertTrue(execute("pollcontent", PollContent.class));
    }

    @Test
    public void test_PollOptionContent() throws FileNotFoundException {
        assertTrue(execute("polloptioncontent", PollOptionContent.class));
    }

    @Test
    public void test_PostActivityTarget() throws FileNotFoundException {
        assertTrue(execute("postactivitytarget", PostActivityTarget.class));
    }
    @Test
    public void test_PurchaseData() throws FileNotFoundException {
        assertTrue(execute("purchasedata", PurchaseData.class));
    }
    @Test
    public void test_ReactionsQuery() throws FileNotFoundException {
        assertTrue(execute("reactionsquery", ReactionsQuery.class));
    }
    @Test
    public void test_RemoveActivitiesQuery() throws FileNotFoundException {
        assertTrue(execute("removeactivitiesquery", RemoveActivitiesQuery.class));
    }
    @Test
    public void test_RemoveGroupMembersQuery() throws FileNotFoundException {
        assertTrue(execute("removegroupmembersquery", RemoveGroupMembersQuery.class));
    }
    @Test
    public void test_TagsQuery() throws FileNotFoundException {
        assertTrue(execute("tagsquery", TagsQuery.class));
    }
    @Test
    public void test_TopicsQuery() throws FileNotFoundException {
        assertTrue(execute("topicsquery", TopicsQuery.class));
    }
    @Test
    public void test_UsersQuery() throws FileNotFoundException {
        assertTrue(execute("usersquery", UsersQuery.class));
    }
    @Test
    public void test_UserId() throws FileNotFoundException {
        assertTrue(execute("userid", UserId.class));
    }
    @Test
    public void test_UserIdList() throws FileNotFoundException {
        assertTrue(execute("useridlist", UserIdList.class));
    }
    @Test
    public void test_UserUpdate() throws FileNotFoundException {
        assertTrue(execute("userupdate", UserUpdate.class));
    }
    @Test
    public void test_InviteContent() throws FileNotFoundException {
        assertTrue(execute("invitecontent", InviteContent.class));
    }
    @Test
    public void test_NotificationBadge() throws FileNotFoundException {
        assertTrue(execute("notificationbadge", NotificationContent.Badge.class));
    }
    @Test
    public void test_NotificationCustomization() throws FileNotFoundException {
        assertTrue(execute("notificationcustomization", NotificationCustomization.class));
    }
    @Test
    public void test_NotificationsQuery() throws FileNotFoundException {
        assertTrue(execute("notificationsquery", NotificationsQuery.class));
    }
    @Test
    public void test_SendNotificationTarget() throws FileNotFoundException {
        assertTrue(execute("sendnotificationtarget", SendNotificationTarget.class));
    }
    @Test
    public void test_ReferralUsersQuery() throws FileNotFoundException {
        assertTrue(execute("referralusersquery", ReferralUsersQuery.class));
    }

    @Test
    public void test_PromoCodeContent() throws FileNotFoundException {
        assertTrue(execute("promocodecontent", PromoCodeContent.class));
    }

    @Test
    public void test_VotesQuery() throws FileNotFoundException {
        assertTrue(execute("votesquery", VotesQuery.class));
    }

    // region Private
    private <T> boolean execute(String directory, Class<T> clazz) throws FileNotFoundException {
        File[] testFiles = getTestFiles(directory);
        if (testFiles.length == 0) {
            throw new FileNotFoundException();
        }
        boolean isValid = false;
        for (File file: testFiles) {
            String testData = getTestFileContent(file);
            isValid = GetSocialJson.parse(testData, clazz) != null;
        }
        return isValid;
    }

    private File[] getTestFiles(String directory) throws FileNotFoundException {
        File file = new File("./../../../jsonbridge-testdata/wrapper_to_native/" + directory);
        if (file.isDirectory()) {
            return file.listFiles();
        }
        throw new FileNotFoundException();

    }
    private String getTestFileContent(File file) throws FileNotFoundException {
        String fileContent = "";
        if (file.exists()) {
            FileInputStream fin = null;
            byte[] byteContent = null;
            try {
                fin = new FileInputStream(file);
                byteContent = new byte[(int) file.length()];
                fin.read(byteContent);
            } catch (Exception ioe) {
                throw new FileNotFoundException();
            } finally {
                // close the streams using close method
                try {
                    if (fin != null) {
                        fin.close();
                    }
                } catch (IOException ioe) {
                    System.out.println("Error while closing stream: " + ioe);
                }
            }
            fileContent = new String(byteContent);
        }
        return fileContent;
    }
}
